File: node/packages/botas/src/turn-context.ts
import type { CoreActivity, ResourceResponse, ChannelAccount } from './core-activity.js'
import type { BotApplication } from './bot-application.js'

const EMOJI_MAP: Record<string, string> = {
  '👍': 'like',
  '❤️': 'heart',
  '😂': 'laugh',
  '😮': 'surprised',
  '😢': 'sad',
  '😠': 'angry'
}

export class TurnContext {
  constructor(
    public readonly app: BotApplication,
    public readonly activity: CoreActivity
  ) {}

  async send(textOrActivity: string | Partial<CoreActivity>): Promise<ResourceResponse | undefined> {
    const activity: Partial<CoreActivity> = 
      typeof textOrActivity === 'string' 
        ? { type: 'message', text: textOrActivity } 
        : { ...textOrActivity }

    // Copy routing fields
    if (!activity.serviceUrl) activity.serviceUrl = this.activity.serviceUrl
    if (!activity.conversation) activity.conversation = { ...this.activity.conversation }
    if (!activity.from) activity.from = { ...this.activity.recipient }
    if (!activity.recipient) activity.recipient = { ...this.activity.from }
    if (!activity.channelId) activity.channelId = this.activity.channelId

    return await this.app.conversationClient.sendCoreActivityAsync(
      activity.serviceUrl!,
      activity.conversation!.id,
      activity
    )
  }

  async sendTyping(): Promise<void> {
    await this.send({ type: 'typing' })
  }

  async sendTargeted(text: string, recipient: ChannelAccount): Promise<ResourceResponse | undefined> {
    return await this.send({
      type: 'message',
      text,
      recipient,
      isTargeted: true
    })
  }

  async addReaction(emojiOrType: string): Promise<void> {
    const reactionType = EMOJI_MAP[emojiOrType] ?? emojiOrType
    if (!this.activity.id) {
      throw new Error('Cannot add reaction: incoming activity has no id')
    }
    await this.app.conversationClient.addReaction(
      this.activity.serviceUrl,
      this.activity.conversation.id,
      this.activity.id,
      reactionType
    )
  }
}

File: dotnet/src/Botas/TurnContext.cs
namespace Botas;

public class TurnContext
{
    private static readonly Dictionary<string, string> EmojiMap = new()
    {
        { "👍", "like" },
        { "❤️", "heart" },
        { "😂", "laugh" },
        { "😮", "surprised" },
        { "😢", "sad" },
        { "😠", "angry" }
    };

    public BotApplication App { get; }
    public CoreActivity Activity { get; }

    public TurnContext(BotApplication app, CoreActivity activity)
    {
        App = app;
        Activity = activity;
    }

    public Task<ResourceResponse> SendAsync(string text, CancellationToken ct = default)
    {
        return SendAsync(new CoreActivity { Type = "message", Text = text }, ct);
    }

    public async Task<ResourceResponse> SendAsync(CoreActivity reply, CancellationToken ct = default)
    {
        if (string.IsNullOrEmpty(reply.ServiceUrl)) reply.ServiceUrl = Activity.ServiceUrl;
        if (string.IsNullOrEmpty(reply.Conversation.Id)) reply.Conversation = new Conversation { Id = Activity.Conversation.Id };
        if (string.IsNullOrEmpty(reply.From.Id)) reply.From = new ChannelAccount { Id = Activity.Recipient.Id, Name = Activity.Recipient.Name };
        if (string.IsNullOrEmpty(reply.Recipient.Id)) reply.Recipient = new ChannelAccount { Id = Activity.From.Id, Name = Activity.From.Name };
        if (string.IsNullOrEmpty(reply.ChannelId)) reply.ChannelId = Activity.ChannelId;

        return await App.ConversationClient.SendCoreActivityAsync(reply, ct);
    }

    public Task SendTypingAsync(CancellationToken ct = default)
    {
        return SendAsync(new CoreActivity { Type = "typing" }, ct);
    }

    public Task<ResourceResponse> SendTargetedAsync(string text, ChannelAccount recipient, CancellationToken ct = default)
    {
        return SendAsync(new CoreActivity
        {
            Type = "message",
            Text = text,
            Recipient = recipient,
            IsTargeted = true
        }, ct);
    }

    public async Task AddReactionAsync(string emojiOrType, CancellationToken ct = default)
    {
        var reactionType = EmojiMap.TryGetValue(emojiOrType, out var mapped) ? mapped : emojiOrType;
        if (string.IsNullOrEmpty(Activity.Id))
        {
            throw new InvalidOperationException("Cannot add reaction: incoming activity has no id");
        }
        await App.ConversationClient.AddReactionAsync(
            Activity.ServiceUrl,
            Activity.Conversation.Id,
            Activity.Id,
            reactionType,
            ct);
    }
}
